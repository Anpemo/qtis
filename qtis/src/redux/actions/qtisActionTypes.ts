export default {
  USER_REGISTER: 'USER_REGISTER',
  PRODUCTS_LIST: 'PRODUCTS_LIST',
  SINGLE_PRODUCT: 'SINGLE_PRODUCT',
  CREATE_PRODUCT: 'CREATE_PRODUCT',
  REVIEWS_LIST: 'REVIEWS_LIST'
}
