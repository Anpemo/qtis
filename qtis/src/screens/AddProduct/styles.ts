
import { COLORS, SIZES, BORDER } from '../../../constants'
import { StyleSheet } from 'react-native'
const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: COLORS.white
  },
  backgroundImage: {
    flex: 1,
    resizeMode: 'cover',
    width: '100%'
  },
  backIcon: {
    color: COLORS.black,
    marginBottom: 5,
    marginLeft: 10,
    position: 'absolute',
    top: 10
  },
  pictureBox: {
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
    height: 200,
    width: 200,
    borderRadius: 100,
    backgroundColor: COLORS.pictureBoxBackground,
    marginBottom: 10
  },
  productPicture: {
    resizeMode: 'contain',
    height: 160,
    width: 160,
    borderRadius: 100,
    backgroundColor: COLORS.white
  },
  inputsContainer: {
    width: '85%',
    height: '50%',
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: SIZES.squareRadius,
    backgroundColor: COLORS.categoryTextBackground
  },
  input: {
    flexGrow: 1,
    width: '90%',
    height: '18%',
    margin: 5,
    justifyContent: 'center',
    backgroundColor: COLORS.cream,
    fontSize: SIZES.p20,
    color: COLORS.lettersGrey,
    fontFamily: 'Montserrat',
    padding: 10
  },
  categoriesInput: {
    flexGrow: 1,
    width: '90%',
    height: '25%',
    margin: 5,
    justifyContent: 'center',
    backgroundColor: COLORS.cream,
    fontSize: SIZES.p20,
    color: COLORS.lettersGrey,
    fontFamily: 'Montserrat',
    padding: 10
  },

  categoriesTitle: {
    fontSize: SIZES.p18,
    color: COLORS.lettersGrey,
    fontFamily: 'Montserrat'
  },
  categoriesModule: {
    backgroundColor: COLORS.white,
    position: 'absolute',
    width: '100%',
    top: -250
  },
  categoryText: {
    fontSize: SIZES.p18,
    fontFamily: 'Montserrat'
  },
  categoryButton: {
    ...BORDER,
    height: 50
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'center'
  },
  button: {
    alignSelf: 'center',
    width: '40%',
    justifyContent: 'center',
    backgroundColor: COLORS.black,
    margin: 5,
    height: SIZES.buttonheight,
    borderRadius: SIZES.buttonRadius
  },
  buttonText: {
    color: COLORS.white,
    textAlign: 'center',
    fontSize: SIZES.buttonText * 0.8,
    fontFamily: 'Montserrat'
  }

})

export default styles
