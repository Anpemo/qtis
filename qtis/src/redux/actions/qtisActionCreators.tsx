import qtisActionTypes from './qtisActionTypes'
import axios from 'axios'
import serverUrls from '../../../constants/serverUrls'
import UserInterface from '../../Interfaces/UserInterface'
import ProductInterface from '../../Interfaces/ProductInterface'
import ReviewInterface from '../../Interfaces/ReviewInterface'
import { Dispatch } from 'react'
import { userRegisterAction, updateUserAction, createReviewAction, userLoginAction, fetchProductsAction, fetchProductAction, fetchReviewsAction, createProductAction } from '../../models'

export function userRegister (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userRegisterAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_REGISTER, userData)

      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_REGISTER,
        data: 401
      })
    }
  }
}

export function userLogin (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<userLoginAction>) {
    try {
      const { data } = await axios.post(serverUrls.BACKEND_LOGIN, userData)
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data
      })
    } catch {
      dispatch({
        type: qtisActionTypes.USER_LOGIN,
        data: 401
      })
    }
  }
}

export function fetchProducts (category: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductsAction>) {
    const { data } = await axios.get(`${serverUrls.PRODUCTS}${category}`)
    dispatch({
      type: qtisActionTypes.PRODUCTS_LIST,
      data
    })
  }
}

export function fetchProduct (barCodeData: String) {
  return async function fetchInfo (dispatch: Dispatch<fetchProductAction>) {
    let date
    try {
      const { data, headers } = await axios.get(`${serverUrls.PRODUCTS}${barCodeData}`)
      date = headers.date
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: data[0]
      })
    } catch {
      dispatch({
        type: qtisActionTypes.SINGLE_PRODUCT,
        data: { data: 'Product does not exist', date }
      })
    }
  }
}

export function createProduct (productData: ProductInterface) {
  return async function fetchInfo (dispatch: Dispatch<createProductAction>) {
    const { data } = await axios.post(serverUrls.PRODUCTS, productData)
    dispatch({
      type: qtisActionTypes.CREATE_PRODUCT,
      data
    })
  }
}

export function cleanProduct () {
  return ({
    type: qtisActionTypes.CLEAN_PRODUCT,
    data: null
  })
}

export function fetchReviews (parameter: Object) {
  return async function fetchInfo (dispatch: Dispatch<fetchReviewsAction>) {
    const { data } = await axios.get(`${serverUrls.REVIEWS}${parameter}`)

    dispatch({
      type: qtisActionTypes.REVIEWS_LIST,
      data
    })
  }
}

export function createReview (reviewData: ReviewInterface) {
  return async function fetchInfo (dispatch: Dispatch<createReviewAction>) {
    const { data } = await axios.post(serverUrls.REVIEWS, reviewData)
    dispatch({
      type: qtisActionTypes.CREATE_REVIEW,
      data
    })
  }
}

export function updateUser (userData: UserInterface) {
  return async function fetchInfo (dispatch: Dispatch<updateUserAction>) {
    const { data } = await axios.put(serverUrls.UPDATE_USER, userData)
    dispatch({
      type: qtisActionTypes.UPDATE_USER,
      data
    })
  }
}
