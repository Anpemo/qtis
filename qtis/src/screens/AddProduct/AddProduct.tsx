import React, { useState } from 'react'
import {
  Image, TouchableOpacity, View, Text, ImageBackground, KeyboardAvoidingView
} from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { images } from '../../../constants'
import styles from './AddProductStyles'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { createProduct } from '../../redux/actions/qtisActionCreators'
import { AntDesign } from '@expo/vector-icons'
import { TextInput } from 'react-native-gesture-handler'
import categories from '../../../constants/categories'

function ProductDetail ({ navigation, route, actions }: any) {
  const [productName, setProductName] = useState('')
  const [brandName, setBrandName] = useState('')
  const [price, setPrice] = useState('')
  const [category, setCategory] = useState('')
  const [openCategories, setOpenCategories] = useState(false)
  const { productBarCode } = route.params

  function selectCategory (category : any) {
    setCategory(category)
    setOpenCategories(false)
  }
  function shareProduct () {
    actions.createProduct({ productBarCode, productName, brandName, price, category })
    navigation.navigate('ProductDetail', { productBarCode })
  }

  return (
    <SafeAreaView style={styles.container}>
      <ImageBackground source={images.backgroundReview} style={styles.backgroundImage}>
        <KeyboardAvoidingView behavior={'height'}>
      <TouchableOpacity onPress={() => { navigation.goBack() }} >
        <AntDesign name="doubleleft" size={22}/>
      </TouchableOpacity>
      <View style={styles.pictureBox}>
            <Image
            source={{ uri: 'https://www.laroche-posay.es/-/media/project/loreal/brand-sites/lrp/emea/es/products/effaclar/effaclar-cleansing-foaming-gel/la-roche-posay-face-cleanser-effaclar-cleansing-foaming-gel-200ml-3337872411083-front.png' }}
            style={styles.productPicture} key={1}
            />
      </View>
      <View style={styles.inputsContainer}>
      <TextInput
          onChangeText={(text) => setProductName(text)}
          placeholder={'Product Name'}
          style={styles.input}
          value={productName}
        />
        <TextInput
          onChangeText={(text) => setBrandName(text)}
          placeholder={'Brand Name'}
          style={styles.input}
          value={brandName}
        />
        <TextInput
          onChangeText={(text) => setPrice(text)}
          placeholder={'Average price'}
          style={styles.input}
          value={price}
        />

      <TouchableOpacity
        onPress={() => {
          setOpenCategories(!openCategories)
        }}
        style={styles.categoriesInput}
        activeOpacity={0.5}
        >
            <View>
              <Text style={styles.categoriesTitle}>{category.toUpperCase() || 'CHOOSE A CATEGORY'}</Text>
              <View style={styles.categoriesModule}>
                {openCategories && categories.map((category) => (
                  <TouchableOpacity
                  key={category.id}
                  activeOpacity={0.5}
                  style={styles.categoryButton}
                  onPress={() => selectCategory(category.name)}
                  >
                    <Text style={styles.categoryText}>{category.name.toUpperCase()}</Text>
                  </TouchableOpacity>
                ))}
              </View>
            </View>

        </TouchableOpacity>

      </View>
      <View style={styles.buttonsContainer}>
        <TouchableOpacity
        onPress={() => shareProduct()}
        style={styles.button}
        >
            <Text style={styles.buttonText}>SHARE</Text>
        </TouchableOpacity>

        <TouchableOpacity
        onPress={() => { navigation.navigate('tabNavigator', { Screen: 'Browser' }) }}
        style={styles.button}
        >
            <Text style={styles.buttonText}>GO TO BROWSER</Text>
        </TouchableOpacity>
      </View>
          </KeyboardAvoidingView>
      </ImageBackground>
    </SafeAreaView>
  )
}
function mapStateToProps ({ productsReducer }: any) {
  return {
    product: productsReducer.product
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    actions: bindActionCreators({
      createProduct
    }, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(ProductDetail)
