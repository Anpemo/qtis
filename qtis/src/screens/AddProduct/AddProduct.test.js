import { cleanup, fireEvent, render, act } from '@testing-library/react-native'
import AddProduct from './AddProduct'
import React from 'react'
import { Provider } from 'react-redux'
import { Alert } from 'react-native'
import configureStore from 'redux-mock-store'
import * as actions from '../../redux/actions/qtisActionCreators'
import * as ImagePicker from 'expo-image-picker'

jest.mock('../../redux/actions/qtisActionCreators')
jest.mock('expo-image-picker')

describe('Given a AddProduct component', () => {
  let component
  const navigate = jest.fn()
  const goBack = jest.fn()
  const params = 123123123

  beforeEach(() => {
    const mockStore = configureStore()

    component = (
          <Provider store={mockStore({
            productReducer: {
              product: {}
            }
          })}><AddProduct navigation={{ navigate, goBack }} route={{ params }} />
          </Provider>
    )
  })

  afterEach(() => {
    cleanup()
  })

  describe('When writing on inputs', () => {
    test('Then productName will change', () => {
      const { getByPlaceholderText } = render(component)
      const productNameInput = getByPlaceholderText('Product Name')
      const newValue = 'newValue'
      act(() => {
        fireEvent.changeText(productNameInput, newValue)
      })
      expect(productNameInput.props.value).toBe(newValue)
    })
    test('Then brandName will change', () => {
      const { getByPlaceholderText } = render(component)

      const brandNameInput = getByPlaceholderText('Brand Name')
      const newValue = 'newValue'
      act(() => {
        fireEvent.changeText(brandNameInput, newValue)
      })
      expect(brandNameInput.props.value).toBe(newValue)
    })
    test('Then price will change', () => {
      const { getByPlaceholderText } = render(component)

      const priceInput = getByPlaceholderText('Average price')
      const newValue = 'newValue'
      act(() => {
        fireEvent.changeText(priceInput, newValue)
      })
      expect(priceInput.props.value).toBe(newValue)
    })
  })
  describe('When pressing on back button', () => {
    test('Then it will navigate', () => {
      const { getByTestId } = render(component)
      const backButton = getByTestId('backButton')
      act(() => {
        fireEvent.press(backButton)
      })
      expect(goBack).toHaveBeenCalled()
    })
  })
  describe('When pressing the SHARE button', () => {
    test('Then createProduct  will be called', () => {
      jest.spyOn(actions, 'createProduct').mockReturnValue({ type: '' })
      const { getByTestId } = render(component)
      const button = getByTestId('shareButton')
      act(() => {
        fireEvent.press(button)
      })
      expect(actions.createProduct).toHaveBeenCalled()
    })
  })
  describe('When pressing the GO TO BROWSER button', () => {
    test('Then navigate  will be called', () => {
      const { getByTestId } = render(component)
      const button = getByTestId('navigatorButton')
      act(() => {
        fireEvent.press(button)
      })
      expect(navigate).toHaveBeenCalled()
    })
  })
  describe('When pressing on categories box', () => {
    test('Then openCategories will be true', () => {
      const selectCategory = jest.fn()
      const { getByTestId, getAllByTestId } = render(component)
      const button = getByTestId('openCategoryButton')
      act(() => {
        fireEvent.press(button)
      })
      const categoryButton = getAllByTestId('categoryButton')[0]
      act(() => {
        fireEvent.press(categoryButton)
      })

      expect(selectCategory).toMatchSnapshot()
    })
  })
  describe('When rendering the component and mocking picker image with cancelled: false ', () => {
    test('Then the button will match the snapshot', async () => {
      const result = {
        cancelled: false,
        uri: 'imageURI'
      }
      jest.spyOn(ImagePicker, 'launchImageLibraryAsync').mockReturnValueOnce(result)

      const { getByTestId } = render(component)
      const button = getByTestId('pickImage')
      act(() => {
        fireEvent.press(button)
      })

      expect(button).toBeTruthy()
    })
  })
  describe('When rendering the component and mocking requestMediaLibraryPermissionsAsync ', () => {
    test('Then alert will be called', async () => {
      const result = {
        status: 'notgranted'
      }
      jest.spyOn(ImagePicker, 'requestMediaLibraryPermissionsAsync').mockReturnValueOnce(result)
      jest.spyOn(Alert, 'alert')

      const { getByText } = render(component)
      act(() => {
        fireEvent.press(getByText('Pick an image from camera roll'))
      })
      expect(Alert.alert).toBeTruthy()
    })
  })
})
