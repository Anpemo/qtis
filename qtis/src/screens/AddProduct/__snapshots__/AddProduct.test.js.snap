// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a AddProduct component When pressing on categories box Then openCategories will be true 1`] = `[MockFunction]`;
